doc/man7/password-handling.pod

```diff
@@ -0,0 +1,95 @@
+=pod
+
+=head1 NAME
+
+password handling - How diverse parts of OpenSSL treat pass phrases
+
+=head1 DESCRIPTION
+
+In a modern world with all sorts of character encodings, the treatment of pass
+phrases has become increasingly complex.
+This manual page attempts to give an overview over how this problem is
+currently addressed in different parts of the openssl library.
+
+=head2 The general case
+
+The OpenSSL library doesn't treat pass phrases in any special way as a general
+rule, and trusts the application or user to choose a suitable character set
+and stick to that throughout the life of affected objects.
+This means that for an object that was encrypted using a pass phrase encoded in
+ISO-8859-1, that object needs to be decrypted using a pass phrase encoded in
+ISO-8859-1.
+Anything else will have the decryption fail.
+
+=head2 PKCS#12
+
+PKCS#12 is a bit different regarding pass phrase encoding.
+The standard stipulates that the pass phrase shall be encoded as a BMPString,
+i.e. UTF-16.
+
+OpenSSL tries to adapt to this requirements in one of the following manners:
+
+=over 4
+
+=item 1.
+
+Tries to read the pass phrase as UTF-8 and re-encode that to UTF-16, or failing
+that, proceeds with step 2.
```

Comment:
It should look at the locale setting instead, and convert from that locale to UTF-16.

On Linux that you get that using nl_langinfo(CODESET), and can the use iconv() to convert it.

On windows, if you're using the *A function, the locale is CP_ACP, and you can convert it using MultiByteToWideChar(). If you're using the *W functions you get it in UCS-2 LE+some additional things, which is close to UTF16, but not the same.

We probably also need to normalize it. Does the standard say which one to use?

Reply:
I don't think the normalisation is specified. ISTR @nmav had an I-D which attempted to do so? I might be making the latter up.

Reply:
Hm, Windows wide chars *do* give you UTF-16 don't they?

Reply:
It seems Windows used to be UCS-2, but since Windows 2000 it's UTF-16

Reply:
And a BMPString is UCS-2 BE.

Reply:
Yes, I had proposed the OpaqueString (RFC7613) normalization
https://tools.ietf.org/html/draft-mavrogiannopoulos-pkcs5-passwords-01

Reply:
We should reject characters that can not be encoded in UCS-2, unless someone updates some standard to say it's UTF-16.

Reply:
For creating files yes. For attempting to read existing files... maybe tolerate UTF-16?

Reply:
That works for me.